protocol/icmpv6.go

```diff
@@ -0,0 +1,643 @@
+	return &MLDQuery{
+		ICMPv6Header: ICMPv6Header{
+			Type: ICMPv6_Type_MLD_Query,
+			Code: 0,
+		},
+		MaxResponse:      maxResponse,
+		MulticastAddress: multicastIP,
+		Version2:         false,
+	}
+}
+
+func NewMLDv2Query(group net.IP, maxResponseTime uint16, queryInterval uint8, sources []net.IP) *MLDQuery {
+	return &MLDQuery{
+		ICMPv6Header: ICMPv6Header{
+			Type: ICMPv6_Type_MLD_Query,
+			Code: 0,
+		},
+		MaxResponse:      maxResponseTime,
+		MulticastAddress: group,
+		Version2:         true,
+		IntervalTime:     queryInterval,
+		NumberOfSources:  uint16(len(sources)),
+		SourceAddresses:  sources,
+	}
+}
+
+//      0                   1                   2                   3
+//     0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
+//    +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+//    |  Type = 143   |    Reserved   |           Checksum            |
+//    +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+//    |           Reserved            |Nr of Mcast Address Records (M)|
+//    +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+//    |                                                               |
+//    .                                                               .
+//    .                  Multicast Address Record [1]                 .
+//    .                                                               .
+//    |                                                               |
+//    +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+//    |                                                               |
+//    .                                                               .
+//    .                  Multicast Address Record [2]                 .
+//    .                                                               .
+//    |                                                               |
+//    +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+//    |                               .                               |
+//    .                               .                               .
+//    |                               .                               |
+//    +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+//    |                                                               |
+//    .                                                               .
+//    .                  Multicast Address Record [M]                 .
+//    .                                                               .
+//    |                                                               |
+//    +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+type MLDv2Report struct {
+	ICMPv6Header
+	Reserved2       uint16
+	NumberOfRecords uint16
+	GroupRecords    []MLDv2Record
+}
+
+func (r *MLDv2Report) Len() (n uint16) {
+	n = uint16(8)
+	for _, r := range r.GroupRecords {
+		n += r.Len()
```

Comment:
It looks confusing to override the method's receiver `r` in the loop? Could we use `gr` as the iterator to avoid it?

Reply:
updated.